### PR TITLE
Add "Tag" parameter to FindAzureResourceCmdlet.

### PR DESCRIPTION
**Pull request by maphad**
Change parameter sets.

Change Parameter Set.

Add comments.

Disallow combining other filters with tag filters as the service does not
support it.

Address code review comments.

Address code review comments.

Record Scenario Test for find resource by tag.

Update command help.

Update help.

Add new file to csproj.

Correct output settings for the json file for the new test.

## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@maphad__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (maphad). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/FindAzureResourceCmdlet.cs [hunk 1]
```diff
@@ -110,19 +125,20 @@ public sealed class FindAzureResourceCmdlet : ResourceManagerCmdletBase
...
         public string ODataQuery { get; set; }
 
+        [Parameter(ParameterSetName = FindAzureResourceCmdlet.ListResourcesByTagObjectParameterSet, Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The tag filter for the OData query. The expected format is @{tagName=$null} or @{tagName = 'tagValue'}.")]
+        public Hashtable Tag { get; set; }
+
         /// <summary>
         /// Gets or sets the tag name.
         /// </summary>
-        [Parameter(ParameterSetName = FindAzureResourceCmdlet.ListResourcesParameterSet, Mandatory = false, HelpMessage = "The name of the tag to query by.")]
-        [Parameter(ParameterSetName = FindAzureResourceCmdlet.MultiSubscriptionListResourcesParameterSet, Mandatory = false, HelpMessage = "The name of the tag to query by.")]
```
Comment (cormacpayne) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/FindAzureResourceCmdlet.cs]: @maphad this is a breaking change - existing scripts that use either of these parameter sets with the `TagName` parameter will no longer work

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/FindAzureResourceCmdlet.cs [hunk 2]
```diff
@@ -110,19 +125,20 @@ public sealed class FindAzureResourceCmdlet : ResourceManagerCmdletBase
...
+        [Parameter(ParameterSetName = FindAzureResourceCmdlet.ListResourcesByTagNameValueParameterSet, Mandatory = false, HelpMessage = "The name of the tag to query by.")]
         [ValidateNotNullOrEmpty]
         public string TagName { get; set; }
 
         /// <summary>
         /// Gets or sets the tag value.
         /// </summary>
-        [Parameter(ParameterSetName = FindAzureResourceCmdlet.ListResourcesParameterSet, Mandatory = false, HelpMessage = "The value of the tag to query by.")]
-        [Parameter(ParameterSetName = FindAzureResourceCmdlet.MultiSubscriptionListResourcesParameterSet, Mandatory = false, HelpMessage = "The value of the tag to query by.")]
```
Comment (cormacpayne) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/FindAzureResourceCmdlet.cs]: @maphad this is a breaking change - existing scripts that use either of these parameter sets with the `TagValue` parameter will no longer work

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/FindAzureResourceCmdlet.cs (hunk 1 above)
Comment (maphad) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/FindAzureResourceCmdlet.cs]: I spoke ti Abhijeet about this and I am looking into it.

Comment (maphad) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/FindAzureResourceCmdlet.cs]: Hi Cormac,

  This change is done intentionally to fix a bug that allowed illegal combination of parameters.
So only TagValue and Top should be allowed with TagName.
Only Top should be allowed with the Tag parameter.

Our service does not support other combinations. Today the user is allowed to specify other combinations but it results in errors like below:
1.	Find-AzureRmResource -TagName "TestTag" -TagValue "testVal1"  -ResourceNameContains "abc"
Gives the following error:
Find-AzureRmResource : InvalidFilterInQueryString : Invalid $filter 'tagName EQ 'TestTag' AND tagValue EQ 'testVal1' AND substringof('abc', name)' specified in the query string.

2.	Find-AzureRmResource -TagName "TestTag" -TagValue "testVal1"  -TenantLevel
Gives the following error:
Find-AzureRmResource : Parameter set cannot be resolved using the specified named parameters.

3.	Find-AzureRmResource -Tag @{TestTag="testVal1"} -ResourceType "Microsoft.Web/sites"
Gives the following error:
Find-AzureRmResource : InvalidFilterInQueryString : Invalid $filter 'resourceType EQ 'Microsoft.Web/sites' AND tagName EQ 'TestTag' AND tagValue EQ 'testVal1'' specified in the query string.

With this parameter set change I am attempting to fix this bug.

Comment (maphad) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/FindAzureResourceCmdlet.cs]: Hi Cormac,

  This change is done intentionally to fix a bug that allowed illegal combination of parameters.
So only TagValue and Top should be allowed with TagName.
Only Top should be allowed with the Tag parameter.

Our service does not support other combinations. Today the user is allowed to specify other combinations but it results in errors like below:
1.	Find-AzureRmResource -TagName "TestTag" -TagValue "testVal1"  -ResourceNameContains "abc"
Gives the following error:
Find-AzureRmResource : InvalidFilterInQueryString : Invalid $filter 'tagName EQ 'TestTag' AND tagValue EQ 'testVal1' AND substringof('abc', name)' specified in the query string.

2.	Find-AzureRmResource -TagName "TestTag" -TagValue "testVal1"  -TenantLevel
Gives the following error:
Find-AzureRmResource : Parameter set cannot be resolved using the specified named parameters.

3.	Find-AzureRmResource -Tag @{TestTag="testVal1"} -ResourceType "Microsoft.Web/sites"
Gives the following error:
Find-AzureRmResource : InvalidFilterInQueryString : Invalid $filter 'resourceType EQ 'Microsoft.Web/sites' AND tagName EQ 'TestTag' AND tagValue EQ 'testVal1'' specified in the query string.

With this parameter set change I am attempting to fix this bug.

Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/FindAzureResourceCmdlet.cs]: @cormacpayne The removed parameter sets are actually not valid parameter sets. There should not be any existing script that works with those two parameter sets, so this change would not break anything. The only concern is that users with the same parameter combinations would now see a different exception.

**Comment by TianoMS** [comment]
PR looks good to me.

**Comment by cormacpayne** [comment]
@maphad Since any combination of `TagName` and `TagValue` used in either `ListResourcesParameterSet` or `MultiSubscriptionListResourcesParameterSet` parameter sets would've resulted in an exception being thrown, we will allow this change to go through.

The only thing that is left to do is to [fill out the change log](https://github.com/Azure/azure-powershell/blob/dev/src/ResourceManager/Resources/ChangeLog.md) under the **Current Release** section with the changes made in this PR

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/FindAzureResourceCmdlet.cs (hunk 1 above)
Comment (maphad) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/FindAzureResourceCmdlet.cs]: I have updated the change log. Thanks for taking a look!

**Pull request merged**